### PR TITLE
common: assert debug mutex lock is not held if !recursive

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
Tested with patch

```
diff --git a/src/osdc/Journaler.cc b/src/osdc/Journaler.cc
index 40cf67702f4c..b80205652d9e 100644
--- a/src/osdc/Journaler.cc
+++ b/src/osdc/Journaler.cc
@@ -1354,7 +1354,7 @@ void Journaler::trim()
 
 void Journaler::_trim()
 {
-  if (state == STATE_STOPPING)
+  if (is_stopping())
     return;
 
   ceph_assert(!readonly);
```

generates

```
   -12> 2024-09-28T01:42:58.131+0000 7f26e0f74640 -1 /home/pdonnell/ceph/src/common/mutex_debug.h: In function 'void ceph::mutex_debug_detail::mutex_debug_impl<<anonymous> >::lock(bool) [with bool Recursive = false]' thread 7f26e0f74640 time 2024-09-28T01:42:58.129910+0000
/home/pdonnell/ceph/src/common/mutex_debug.h: 177: FAILED ceph_assert(recursive || !is_locked_by_me())

 ceph version 19.3.0-5270-g15c2fbe0f951 (15c2fbe0f9516c3b633f3bd40d80e52e56640b17) squid (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x11c) [0x7f26ed4c6bc9]
 2: (ceph::register_assert_context(ceph::common::CephContext*)+0) [0x7f26ed4c6dd0]
 3: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool)+0x81) [0x55f135e35919]
 4: (Journaler::_trim()+0x3e) [0x55f1362f7056]
 5: (Journaler::_finish_write_head(int, Journaler::Header&, C_OnFinisher*)+0x584) [0x55f1362fa592]
 6: (Journaler::C_WriteHead::finish(int)+0x1b) [0x55f1363048ed]
 7: (Context::complete(int)+0x9) [0x55f135e4c645]
 8: (Finisher::finisher_thread_entry()+0x5e1) [0x7f26ed44ea2b]
 9: (Finisher::FinisherThread::entry()+0xd) [0x55f135e8220f]
 10: (Thread::entry_wrapper()+0x3f) [0x7f26ed494ae7]
 11: (Thread::_entry_func(void*)+0x9) [0x7f26ed494aff]
 12: /lib64/libc.so.6(+0x89c52) [0x7f26ec289c52]
 13: /lib64/libc.so.6(+0x10ec80) [0x7f26ec30ec80]
```

seems to work as expected.

**Comment by batrick** [comment]
```
[ RUN      ] MutexDebug.NotRecursive
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/mutex_debug.h: In function 'void ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool) [Recursive = false]' thread 7f75383fe440 time 2024-09-28T02:36:04.468287+0000
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/mutex_debug.h: 177: FAILED ceph_assert(recursive || !is_locked_by_me())
 ceph version Development (no_version) squid (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x1b4) [0x7f753b2c3674]
 2: /home/jenkins-build/build/workspace/ceph-pull-requests/build/lib/libceph-common.so.2(+0x16e74bf) [0x7f753b2c34bf]
 3: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool)+0x3e) [0x559467bd536e]
 4: (MutexDebug_NotRecursive_Test::TestBody()+0x93c) [0x559467bd09dc]
 5: (void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x7b) [0x559467c42e0b]
 6: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x6a) [0x559467c26b9a]
 7: (testing::Test::Run()+0xc3) [0x559467c048a3]
 8: (testing::TestInfo::Run()+0xda) [0x559467c055ea]
 9: (testing::TestSuite::Run()+0xfb) [0x559467c05e0b]
 10: (testing::internal::UnitTestImpl::RunAllTests()+0x442) [0x559467c13522]
 11: (bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x7b) [0x559467c45edb]
 12: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x6a) [0x559467c2949a]
 13: (testing::UnitTest::Run()+0xcb) [0x559467c1308b]
 14: (RUN_ALL_TESTS()+0x11) [0x559467be2911]
 15: main()
 16: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7f7538fc0d90]
 17: __libc_start_main()
 18: _start()
```

fun. moving to draft.

**Review by rzarzynski — APPROVED** [review]

File: src/common/mutex_debug.h [hunk 1]
```diff
@@ -198,6 +194,16 @@ class mutex_debug_impl : public mutex_debugging_base
     unlock_impl();
   }
 
+private:
+  bool _try_lock(bool no_lockdep = false) {
```
Comment (idryomov) [inline src/common/mutex_debug.h:198]: Being an internal helper, does `_try_lock()` need to default the argument?
```suggestion
  bool _try_lock(bool no_lockdep) {
```

File: src/test/common/test_mutex_debug.cc [hunk 2]
```diff
@@ -65,7 +65,6 @@ TEST(MutexDebug, NotRecursive) {
   ASSERT_TRUE(m.is_locked());
   ASSERT_FALSE(std::async(std::launch::async, ttl, &m).get());
 
-  ASSERT_THROW(m.lock(), std::system_error);
```
Comment (idryomov) [inline src/test/common/test_mutex_debug.cc:68]: > common/test: do not test exception raised from recursive lock 
>
> [...] Our implementation of debug_mutex now catches this error with a ceph_assert so it cannot be caught.

Adding a death test might be appropriate.

File: src/common/mutex_debug.h (hunk 1 above)
Comment (batrick) [reply src/common/mutex_debug.h]: done

File: src/test/common/test_mutex_debug.cc [hunk 3]
```diff
@@ -57,6 +57,14 @@ TEST(MutexDebug, Lock) {
   test_lock<ceph::mutex_debug>();
 }
 
+TEST(MutexDebug, DeathTestNotRecursive) {
```
Comment (cbodley) [inline src/test/common/test_mutex_debug.cc:60]: https://google.github.io/googletest/advanced.html#death-tests-and-threads recommends adding "DeathTest" to the end of the suite name (ie `MutexDebugDeathTest`) so it knows to run those before any other test cases spawn threads

Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: ah, thanks

Comment (cbodley) [reply src/test/common/test_mutex_debug.cc]: that was about the _suite_ name, not the test name
```diff
-TEST(MutexDebug, NotRecursiveDeathTest) {
+TEST(MutexDebugDeathTest, NotRecursive) {
```

Comment (batrick) [reply src/test/common/test_mutex_debug.cc]: Right, thanks!

**Comment by batrick** [comment]
jenkins test make check arm64

File: src/test/common/test_mutex_debug.cc [hunk 4]
```diff
@@ -57,6 +57,14 @@ TEST(MutexDebug, Lock) {
   test_lock<ceph::mutex_debug>();
 }
 
+TEST(MutexDebugDeathTest, NotRecursive) {
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: I was trying to keep the test extremely specific / minimalistic. The test I pulled it out of was really testing other things (confusingly). Perhaps the remaining NotRecursive test should be simply removed.

File: src/test/common/test_mutex_debug.cc [hunk 5]
```diff
@@ -57,6 +57,14 @@ TEST(MutexDebug, Lock) {
...
+  // avoid assert during test cleanup where the mutex is locked and cannot be pthread_mutex_destroy'd
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:61]: ok

File: src/test/common/test_mutex_debug.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: I just removed the extraneous parts of that test.

[the remark this answers is not in the document]
Comment (idryomov) [reply src/test/common/test_mutex_debug.cc:60]: I'm not sure I agree with those parts being extraneous.  What would test `try_lock()` on `mutex_debug` variant with them removed?

[the remark this answers is not in the document]
Comment (idryomov) [reply src/test/common/test_mutex_debug.cc:60]: `ASSERT_DEATH` doesn't need to be the last statement in the test, so just sticking `ASSERT_DEATH` before the `unlock()` call and keeping everything else in place should work just fine.

[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: > I'm not sure I agree with those parts being extraneous. What would test `try_lock()` on `mutex_debug` variant with them removed?

A general guideline for unit tests is that they be as short and specific as possible in what they test. There is no need to launch threads which attempt to lock the mutex in order to verify that double-locking a `ceph::mutex_debug` will result in an abort.

If we want to verify that the mutex actually "works" (ya, what a can of worms that is) with multiple threads then it should be a separate test.

[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: > `ASSERT_DEATH` doesn't need to be the last statement in the test, so just sticking `ASSERT_DEATH` before the call to `unlock()` and keeping everything else in place should work.

I didn't realize that at first but I don't think that changes my opinion about the test.

[the remark this answers is not in the document]
Comment (batrick) [reply src/test/common/test_mutex_debug.cc:60]: Regarding `try_lock()`, I think that's already tested in `test_lock`, no?

[the remark this answers is not in the document]
Comment (idryomov) [reply src/test/common/test_mutex_debug.cc:60]: > Regarding `try_lock()`, I think that's already tested in `test_lock`, no?

Yeah, it looks like it.  I missed that it's a template that is instantiated both on `mutex_debug` and `mutex_recursive_debug`, sorry.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
jenkins test make check

**Comment by batrick** [comment]
```
FATAL: Channel "hudson.remoting.Channel@2132300:JNLP4-connect connection from 8.43.84.3/8.43.84.3:52671": Remote call on JNLP4-connect connection from 8.43.84.3/8.43.84.3:52671 failed. The channel is closing down or has closed down
java.nio.channels.ClosedChannelException
	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:155)
	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:143)
	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:789)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at jenkins.util.ErrorLoggingExecutorService.lambda$wrap$0(ErrorLoggingExecutorService.java:51)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@2132300:JNLP4-connect connection from 8.43.84.3/8.43.84.3:52671": Remote call on JNLP4-connect connection from 8.43.84.3/8.43.84.3:52671 failed. The channel is closing down or has closed down
	at hudson.remoting.Channel.call(Channel.java:996)
	at hudson.Launcher$RemoteLauncher.kill(Launcher.java:1147)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:538)
	at hudson.model.Run.execute(Run.java:1894)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:44)
	at hudson.model.ResourceController.execute(ResourceController.java:101)
	at hudson.model.Executor.run(Executor.java:442)
```

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
https://tracker.ceph.com/issues/68391#note-1

**Pull request merged**